src/tt.cpp

```diff
@@ -28,30 +28,202 @@
 #include "tt.h"
 #include "uci.h"
 
+namespace {
+
+inline void refreshGen5(TTEntryPacked &tte, uint8_t newGen5)
+{
+  setBitField<TTEntryPacked::Gen5>(tte.bits, newGen5);
+}
+
+inline int32_t ageDepthByGen(uint8_t depth8, uint8_t curGen5, uint8_t prevGen5)
+{
+  constexpr uint8_t genDepthPenalty = 8; // every gen means depth reduction by 8
+  const uint8_t aging = ((curGen5 - prevGen5) & 0x1FU) * genDepthPenalty;
+
+  return int32_t(depth8) - int32_t(aging);
+}
+
+inline uint32_t getClusterEntryKey(uint64_t clusterKeys, uint8_t slotIndex)
+{
+  constexpr unsigned fieldBits = TranspositionTable::HashEntryKeyField::numBits;;
+  return bitExtract(clusterKeys, slotIndex * fieldBits, fieldBits);
+}
+
+inline void setClusterEntryKey(uint64_t &clusterKeys, uint8_t slotIndex, uint32_t newKey)
+{
+  constexpr unsigned fieldBits = TranspositionTable::HashEntryKeyField::numBits;;
+  clusterKeys = bitFieldSet(clusterKeys, slotIndex * fieldBits, fieldBits, newKey);
+}
+
+inline uint32_t entryKeyFromZobrist(Key k)
+{
+  return extractBitField<TranspositionTable::HashEntryKeyField>(k);
+}
+
+inline uint32_t extraKeyFromZobrist(Key k)
+{
+  return extractBitField<TranspositionTable::HashExtraKeyField>(k);
+}
+
+Move decodeTTMove(uint32_t encodedMove)
+{
+  encodedMove -= static_cast<bool>(encodedMove);
+
+  if (encodedMove & 0x1000U)
+  {
+      // special move type, need to figure out which one by from source rank
+
+      uint32_t ret;
+      uint32_t promotionBits;
+
+      encodedMove &= 0xFFFU;
+
+      switch ((encodedMove & 07000U) >> 9)
+      {
+          case RANK_1: // 0, 7
+          case RANK_8:
+              ret = encodedMove | CASTLING;
+              break;
+
+          case RANK_2: // black promotion
+              promotionBits = PROMOTION | 00000U | (encodedMove & 00070) << 9;
+              ret = (encodedMove & 07707U) | promotionBits;
+              break;
+
+          case RANK_7: // write promotion
+              promotionBits = PROMOTION | 00070U | (encodedMove & 00070) << 9;
+              ret = (encodedMove & 07707U) | promotionBits;
+              break;
+
+              // rank 4/5
+          default:
+              ret = ENPASSANT | encodedMove;
+              break;
```

Comment:
If this encoding scheme is kept this could maybe be an array `or_type[RANK_NB]` so this can be done as `encodedMove |= or_type[sq_rank]` and just special case `RANK_2` and `RANK_7`. 

For `case RANK_7:`  `& 07707U` seems redundant because it will always be ored with `00070U` and is already `&= 0xFFF`?